Remove @IntermediateInputFunction annotation

maciejgrzybek opened the pull request:
@IntermediateInputFunction is a facilitation when `combine` function is identical to `input` function. This facilitation is rather small and applicable in a few cases only but it adds a burden of inconsistency. Moreover, it's quite misleading to have two different approaches to the same problem (combine method).

Commit-by-commit review suggested.

maciejgrzybek commented:
CC: @fiedukow, @arhimondr

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/DoubleSumAggregation.java:
```diff
@@ -25,13 +25,29 @@
+            if (otherState.isNull()) {
+                return;
+            }
+            state.setNull(false);
```
why not just remove the condition above and do `state.setNull(otherState.isNull)`? I think it is more readable and looks the same as similar code for double. Will it throw on calling `getDouble` for the nullable value?

maciejgrzybek replied:
I wanted to avoid getting uninitialized value because IIRC generated c-tor for State classes does not initialize fields.
Maybe @cberner may comment on this?

haozhun replied:
TL;DR `otherState.isNull()` is guaranteed to be false. Therefore you can remove `if`s that is guaranteed to go one way or the other.

There are 2 guarantees provided by the aggregation function framework:
* `otherState` is guaranteed to be the result of `deserialize`.
* When deserializing, `null` values are skipped.

For most states, including `NullableDoubleState`, null states are serialized as `null` value. Therefore, `otherState` cannot be a null state.

I understand that the guarantees provided here are complex. When the aggregation function framework came out, we didn't document them, in the hope that implementations will just do the right thing as if no such guarantee is provided. We ended up failing miserably, having inconsistent and self-contradictory code everywhere for aggregation functions. Yet these seemingly unreasonable code are all correct. (The tests that get run on aggregation functions covers null related corner cases pretty well.)

Up until recently, these guarantees are not documented.

arhimondr commented:
lgtm

electrum commented:
I'd squash the commits "Remove @IntermediateInputFunction" except the last one (that removes the annotation), since they're all basically the same and easy to review together.

maciejgrzybek commented:
@electrum Of course, I won't object. It's just simpler to review smaller commits, then squash rather than having big commits and maybe unsquashing them when needed.

maciejgrzybek commented:
Rebased and resolved conflicts.

maciejgrzybek commented:
I've rewritten commits to be more succinct and more descriptive.
Squashed ed2fae8, 1ec8457, d01274c, 616e8c6 and d615c5c into eab027c.
I left 037a77f (after rebase on most recent master it became 122caae) and 3c9f8b9 (ab71bef) because there was a test missing which I added while refactoring. I wanted to provide test first to make sure it passes before change and after applying patch with refactoring.
I also squashed 9a795e2 and 1f2c8e6 into 3201fe8.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongSumAggregation.java:
```diff
@@ -30,13 +30,29 @@
+    public static void combine(NullableLongState state, NullableLongState otherState)
+    {
+        if (state.isNull()) {
+            if (otherState.isNull()) {
```
ditto

haozhun commented on presto-tests/src/main/java/com/facebook/presto/tests/CustomSum.java:
```diff
@@ -29,13 +29,29 @@
+    public static void combine(NullableLongState state, NullableLongState otherState)
+    {
+        if (state.isNull()) {
+            if (otherState.isNull()) {
```
ditto

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregationFunction.java:
```diff
@@ -256,4 +269,112 @@ public static void input(MethodHandle methodHandle, BlockState state, Block valu
+
+    public static void combine(MethodHandle methodHandle, NullableLongState state, NullableLongState otherState)
+    {
+        if (state.isNull() && !otherState.isNull()) {
```
I would remove tests for `otherState.isNull` for the same reason I mentioned before.

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArbitraryAggregationFunction.java:
```diff
@@ -198,4 +210,47 @@ public static void input(Type type, BlockState state, Block block, int position)
+
+    public static void combine(NullableLongState state, NullableLongState otherState)
+    {
+        if (!state.isNull() || otherState.isNull()) {
```
I would remove all `otherState.isNull()` checks here for the same reason I mentioned before.

haozhun commented:
I have some minor comments.

Everything looks pretty good otherwise. I'll merge after you address the comments. Thank you!

maciejgrzybek commented:
@haozhun: I have addressed your comments. I have amended e212014 and 4debb24 to remove unnecessary `otherState.isNull()` checks.

haozhun commented:
Merged #5700. Thank you.
